In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [2]:
data=pd.read_csv("/kaggle/input/creditcardfraud/creditcard.csv")
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
data.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [4]:
data=data.drop_duplicates()

In [5]:
x=data.drop("Class",axis=1)
y=data["Class"]

In [6]:
xtrain, xtest, ytrain, ytest = train_test_split(x, y, test_size=0.2, random_state=101)

In [7]:
smote = SMOTE(random_state=101)
xtrainsmote, ytrainsmote = smote.fit_resample(xtrain, ytrain)

In [8]:
xtrainsmote.head()

Time        V1        V2        V3        V4        V5        V6  \
0   41110.0  1.360170 -0.105120 -0.026442 -0.246305 -0.375191 -0.726283   
1  117430.0 -0.144263  0.989749 -0.658424 -0.765560  1.044936 -0.408662   
2   30637.0 -0.375379  0.976726  1.340372 -0.141541  0.095680 -0.633842   
3  165836.0  2.275658 -0.625857 -1.747002 -1.164417  0.120648 -0.503048   
4  164283.0 -5.943253  4.895733 -4.256435 -1.647317 -1.172500 -2.153876   

         V7        V8        V9  ...       V20       V21       V22       V23  \
0 -0.163242 -0.081611  0.581652  ... -0.161762 -0.186852 -0.598759 -0.034795   
1  0.707224  0.147879  0.119354  ...  0.069819 -0.362937 -0.922522  0.052104   
2  0.634357  0.014203 -0.231337  ...  0.055395 -0.271365 -0.756375 -0.076125   
3 -0.302738 -0.340158 -0.688982  ...  0.095707  0.401617  1.213436 -0.074032   
4 -0.232372  1.346291  2.520441  ...  0.767148 -0.126537  0.353528  0.339498   

        V24       V25       V26       V27       V28  Amount  
0 -0.444342  0.292029  1.018943 -0.096247 -0.009115    0.69  
1  0.138419 -0.400572  0.126464  0.325634  0.124415    1.79  
2 -0.157236 -0.157060  0.105035  0.259487  0.113986    1.78  
3  0.119684  0.344321  0.076956 -0.033690 -0.063354   15.00  
4 -0.130007  0.961094  0.107010 -0.518102 -0.158214    0.77  

[5 rows x 30 columns]

In [9]:
ytrainsmote.head()

0    0
1    0
2    0
3    0
4    0
Name: Class, dtype: int64

In [10]:
models={"Decision Tree":DecisionTreeClassifier(),
"K Neighbors":KNeighborsClassifier(),
"Random Forest Classifier":RandomForestClassifier()}

In [11]:
for name, model in models.items():
    model.fit(xtrainsmote, ytrainsmote)
    ypredict = model.predict(xtest)
    acc=accuracy_score(ytest,ypredict)
    print(f"\nClassifier: {name}")
    print("-----------------------------------------------------")
    print("Accuracy : ",round(acc*100,2))
    print("Classification Report :")
    print(classification_report(ytest, ypredict))
    print("\n Confusion Matrix :")
    print(confusion_matrix(ytest, ypredict))
    print("=====================================================")


Classifier: Decision Tree
-----------------------------------------------------
Accuracy :  99.78
Classification Report :
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56657
           1       0.40      0.82      0.53        89

    accuracy                           1.00     56746
   macro avg       0.70      0.91      0.77     56746
weighted avg       1.00      1.00      1.00     56746


 Confusion Matrix :
[[56546   111]
 [   16    73]]

Classifier: K Neighbors
-----------------------------------------------------
Accuracy :  94.51
Classification Report :
              precision    recall  f1-score   support

           0       1.00      0.95      0.97     56657
           1       0.01      0.44      0.02        89

    accuracy                           0.95     56746
   macro avg       0.51      0.69      0.50     56746
weighted avg       1.00      0.95      0.97     56746


 Confusion Matrix :
[[53591  3066]
 [   50    39]

# **Conclusion**
> **Our evaluation of various classifiers for predicting credit card fraud demonstrates the superior performance of the Random Forest Classifier. With an accuracy of 99.96%, it outperforms both the Decision Tree and K Neighbors classifiers. The Random Forest Classifier also maintains high precision and recall rates of 0.89 and 0.82, respectively, for detecting frauds, resulting in an impressive f1-score of 0.85. In comparison, while the Decision Tree achieved an accuracy of 99.78%, its ability to detect faults was less effective, with a recall of 0.8 and a lower f1-score of 0.53. The K Neighbors classifier, with an accuracy of 94.51%, struggled significantly with fraud detection, achieving a recall of only 0.44 and an f1-score of 0.02. These results suggest that the *Random Forest Classifier* is the most reliable model for *credit card fraud detection*, offering a balanced performance in terms of accuracy, precision, and recall.**